numpy/core/src/umath/ufunc_object.c

```diff
@@ -2651,195 +2666,118 @@ PyUFunc_GenericFunction(PyUFuncObject *NPY_UNUSED(ufunc),
 
 
 /*
- * Given the output type, finds the specified binary op.  The
- * ufunc must have nin==2 and nout==1.  The function may modify
- * otype if the given type isn't found.
+ * Promote and resolve a reduction like operation.
  *
- * Returns 0 on success, -1 on failure.
+ * @param ufunc
+ * @param arr The operation array (out was never used)
```

Comment:
This is indeed confusing, changed.  What NumPy used to do is do the same as if `dtype=out.dtype` was passed, rather than pass `out` itself as an operand to the type resolution.  I bet there are some subtle differences that barely matter.